browser/brave_app_controller_mac.mm

```diff
@@ -0,0 +1,103 @@
+/* Copyright (c) 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#import "brave/browser/brave_app_controller_mac.h"
+
+#include "brave/app/brave_command_ids.h"
+#include "brave/browser/ui/browser_commands.h"
+#include "chrome/app/chrome_command_ids.h"
+#include "chrome/browser/ui/browser.h"
+#include "chrome/browser/ui/browser_commands.h"
+#include "chrome/browser/ui/browser_finder.h"
+#include "chrome/grit/generated_resources.h"
+
+@implementation BraveAppController
+
+- (void)mainMenuCreated {
+  [super mainMenuCreated];
+
+  NSMenu* editMenu = [[[NSApp mainMenu] itemWithTag:IDC_EDIT_MENU] submenu];
+  _copyMenuItem = [editMenu itemWithTag:IDC_CONTENT_CONTEXT_COPY];
+  DCHECK(_copyMenuItem);
+  [[_copyMenuItem menu] setDelegate:self];
+  _copyCleanLinkMenuItem = [editMenu itemWithTag:IDC_COPY_CLEAN_LINK];
+  DCHECK(_copyCleanLinkMenuItem);
+  [[_copyCleanLinkMenuItem menu] setDelegate:self];
+}
+
+- (void)dealloc {
+  [[_copyMenuItem menu] setDelegate:nil];
+  [[_copyCleanLinkMenuItem menu] setDelegate:nil];
+  [super dealloc];
+}
+
+- (Browser*)getBrowser {
+  return chrome::FindBrowserWithProfile([self lastProfileIfLoaded]);
+}
+
+- (BOOL)shouldShowCleanLinkItem {
+  if (_hasSelectedURLForTesting.has_value()) {
+    return _hasSelectedURLForTesting.value();
+  }
+  return brave::HasSelectedURL([self getBrowser]);
+}
+
+- (void)setKeyEquivalentToItem:(NSMenuItem*)item {
+  auto* hootkeyItem =
+      item == _copyMenuItem ? _copyMenuItem : _copyCleanLinkMenuItem;
+  auto* noHootkeyItem =
+      item == _copyMenuItem ? _copyCleanLinkMenuItem : _copyMenuItem;
+
+  [hootkeyItem setKeyEquivalent:@"c"];
+  [hootkeyItem setKeyEquivalentModifierMask:NSEventModifierFlagCommand];
+
+  [noHootkeyItem setKeyEquivalent:@""];
+  [noHootkeyItem setKeyEquivalentModifierMask:0];
+}
+
+- (void)menuNeedsUpdate:(NSMenu*)menu {
```

Comment:
shouldn't we be overriding `BuildEditMenu` main_menu_builder?

Reply:
I show item when we have url selected in the url bar and hide if not. `BuildEditMenu` called once on the browser launch.

Reply:
Yes, but this is completely different from how every other menu item works. That has several problems:
1. It's fragile against upstream changes
2. It's harder to verify that it works correctly because it's different from everything else
3. It's harder to understand the code
So I'm asking if there's some reason you didn't do this the same way that the rest of the menu items. Also I believe that disabling the item is preferred in the menubar (vs hiding it completely)

Reply:
and I'm not saying to only use `BuildEditMenu` because I understand the menu item has to change depending on context, but other menu items do that as well so I'm asking why you didn't follow the existing patterns for handling menu items

Reply:
There is `_closeTabMenuItem` which is also updated on `menuNeedsUpdate`, I reused existing pattern, @bridiver do you have any other suggestions here?

Reply:
There is `_closeTabMenuItem` which is updated on `menuNeedsUpdate` https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/app_controller_mac.mm;l=774?q=chrome%2Fbrowser%2Fapp_controller_mac.mm&ss=chromium, I reused this existing pattern to work with this new menu item, @bridiver do you have any more detailed suggestions here?

Reply:
I think he did follow the existing pattern in Chromium actually @bridiver. You can check out comments about the `app_controller_mac` here in upstream:
https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/app_controller_mac.h;l=48-55

```
// This handles things like responding to menus when there are no windows
// open, etc and acts as the NSApplication delegate.
```

The windows being open is a good example of a dynamic behavior. The menu item isn't there if all the windows are closed. This "copy clean link" is a similar - the menu item shouldn't be there unless the conditions are correct.

If we didn't care about hiding/showing the menu item and always wanted the menu item to be there (enable/disable depending on the context), we could put a new entry in https://github.com/brave/brave-core/blob/master/chromium_src/chrome/browser/ui/cocoa/main_menu_builder.mm

Something like:
```
#define BRAVE_BUILD_EDIT_MENU      \
  Item(IDS_COPY_CLEAN_LINK)        \
      .command_id(IDC_COPY_CLEAN_LINK),
```

And then place this with patch below the regular `IDS_COPY` in `BuildEditMenu`. But the code in `BuildEditMenu` is never re-evaluated. Unless you use something like the example @spylogsster had, you can only really enable/disable items AFAIK.